core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/StackSyncService.java

```diff
@@ -0,0 +1,141 @@
+            InstanceSyncState state = syncStateFinders.get(stack.cloudPlatform()).getState(stackId, instanceMetaData.getInstanceId());
+            if (DELETED.equals(state) && !instanceMetaData.isTerminated()) {
+                updateMetadataToTerminatedAndDeregisterFromAmbari(stackId, stack, instanceMetaData, instanceGroup);
+            } else if (RUNNING.equals(state) && !instanceMetaData.isRegistered()) {
+                updateMetaDataToRunning(stackId, instanceMetaData, instanceGroup);
+            } else if (STOPPED.equals(state) && !instanceMetaData.isTerminated()) {
+                updateMetaDataToTerminated(stackId, instanceMetaData, instanceGroup);
+            }
+        }
+        if (Status.stopStatusesForUpdate().contains(stack.getStatus())) {
+            stackUpdater.updateStackStatus(stack.getId(), STOPPED, statusReason);
+        } else if (Status.availableStatusesForUpdate().contains(stack.getStatus())) {
+            stackUpdater.updateStackStatus(stack.getId(), AVAILABLE, statusReason);
+        }
+
+
+    }
+
+    private void updateMetadataToTerminatedAndDeregisterFromAmbari(Long stackId, Stack stack, InstanceMetaData instanceMetaData, InstanceGroup instanceGroup) {
+        instanceGroup.setNodeCount(instanceGroup.getNodeCount() - 1);
+        instanceMetaData.setInstanceStatus(TERMINATED);
+        HostGroup hostGroup = hostGroupRepository.findHostGroupsByInstanceGroupName(stack.getCluster().getId(), instanceGroup.getGroupName());
+        HostMetadata data = hostMetadataRepository.findHostsInClusterByName(stack.getCluster().getId(), instanceMetaData.getDiscoveryFQDN());
+        try {
+            deregisterFromAmbari(stack, hostGroup, data);
+            instanceMetaDataRepository.save(instanceMetaData);
+            instanceGroupRepository.save(instanceGroup);
+            eventService.fireCloudbreakEvent(stackId, AVAILABLE.name(),
+                    String.format("Instance %s was terminated by hand update metadata...", instanceMetaData.getInstanceId()));
+        } catch (Exception ex) {
+            LOGGER.error("Terminated instance deregistration from ambari was unsuccess: ", ex);
+        }
+    }
+
+    private void updateMetaDataToTerminated(Long stackId, InstanceMetaData instanceMetaData, InstanceGroup instanceGroup) {
+        instanceGroup.setNodeCount(instanceGroup.getNodeCount() - 1);
+        instanceMetaData.setInstanceStatus(TERMINATED);
+        instanceMetaDataRepository.save(instanceMetaData);
+        instanceGroupRepository.save(instanceGroup);
+        eventService.fireCloudbreakEvent(stackId, AVAILABLE.name(),
+                String.format("Instance %s was stopped by hand update metadata...", instanceMetaData.getInstanceId()));
+    }
+
+    private void updateMetaDataToRunning(Long stackId, InstanceMetaData instanceMetaData, InstanceGroup instanceGroup) {
+        instanceGroup.setNodeCount(instanceGroup.getNodeCount() + 1);
+        instanceMetaData.setInstanceStatus(REGISTERED);
```

Comment:
the same as above: how do we know that this instance is registered? If it was added to the stack only and later restarted, it can be unregistered as well. I guess we should check here if it is known by Ambari.